p2p/simulations/network.go

```diff
@@ -718,17 +721,102 @@ func (net *Network) Load(snap *Snapshot) error {
+	eventLoopStarted := make(chan struct{}) // ensures that event loop is started before it is closed
+	done := make(chan struct{})             // ensures that the event loop goroutine is terminated
+	defer close(done)
+
+	go func() {
+		// Subscribe to event channel.
+		events := make(chan *Event)
+		sub := net.Events().Subscribe(events)
+		defer sub.Unsubscribe()
+
+		// Expected number of connections.
+		total := len(snap.Conns)
+		// Set of all established connections from the snapshot, not other connections.
+		// Key array element 0 is the connection One field value, and element 1 connection Other field.
+		connections := make(map[[2]enode.ID]struct{}, total)
+
+		// once is a closed channel that is read in the event loop below
+		// only once.
+		// It ensures that eventLoopStarted is closed which signals that
+		// it is safe to call connect method on the network without the
+		// possibility to miss a few first connection events.
+		once := make(chan struct{})
+		// Close once channel so that it can be read from in the event loop.
+		close(once)
+
+		for {
+			select {
+			case e := <-events:
+				// Detect only connection events.
+				if e.Type != EventTypeConn {
+					continue
+				}
+				connection := [2]enode.ID{e.Conn.One, e.Conn.Other}
+				// Nodes are still not connected or have been disconnected.
+				if !e.Conn.Up {
+					// Delete the connection from the set of established connections.
+					// This will prevent false positive in case disconnections happen.
+					delete(connections, connection)
+					continue
+				}
+				// Check that the connection is from the snapshot.
+				for _, conn := range snap.Conns {
+					if conn.One == e.Conn.One && conn.Other == e.Conn.Other {
+						// Add the connection to the set of established connections.
+						connections[connection] = struct{}{}
+						if len(connections) == total {
+							// Signal that all nodes are connected.
+							close(allConnected)
+							return
+						}
+
+						break
+					}
+				}
+			case <-once:
+				// Set once to nil as nil channel never blocks forever.
+				// This ensures that this for loop never gets into this part
+				// of the code again.
+				once = nil
+				// Proceed with connecting the nodes, as we are ready to
+				// detect events.
+				close(eventLoopStarted)
+			case <-done:
+				// Load function returned, terminate this goroutine.
+				return
+			}
+		}
+	}()
+
+	// Do not proceed until the goroutine with the event loop actually is ready
+	// to receive events.
+	<-eventLoopStarted
+
+	// Start connecting.
 	for _, conn := range snap.Conns {
 
 		if !net.GetNode(conn.One).Up || !net.GetNode(conn.Other).Up {
 			//in this case, at least one of the nodes of a connection is not up,
 			//so it would result in the snapshot `Load` to fail
 			continue
 		}
-		if err := net.Connect(conn.One, conn.Other); err != nil {
+		if err := net.Connect(conn.One, conn.Other); err != nil && !strings.Contains(err.Error(), "already connected") {
```

Comment:
Why should "already connected" even be possible here?

Reply:
It is possible if discovery is set to true, which is not disabled in snapshot load function, but in the provided service. This change is protecting agains such scenario, which happened quite frequently when discovery is set to true or the fix NotifyPeer is not applied. I would keep ignoring this error as it may result in flaky tests if discovery is set to true.

Reply:
Isn't this in the end a bit of a catch 22 scenario? We want to make sure that the network is _exactly_ as we had it in the snapshot, but if we boot with discovery it will never be. We should never pretend that it should. Maybe this warrants a bit more discussion?

Reply:
I agree that it needs more discussion. My reason for this approach is that the flakiness produced with exposing this error is not contributing to what is the core reason for `already connected` error. By reporting `already connected` error, we are requiring that discovery is set to false before the snapshot is loaded, and I am not sure if it is safe and correct to set it to true in the runtime, after the snapshot is loaded, if it needs to be enabled.

Reply:
Cool, I will revert this change as discovery is a not needed.